lib/Sema/CSBindings.cpp

```diff
@@ -75,13 +83,284 @@ void ConstraintSystem::inferTransitiveSupertypeBindings(
+  llvm::SmallPtrSet<CanType, 4> existingTypes;
+  for (const auto &binding : Bindings)
+    existingTypes.insert(binding.BindingType->getCanonicalType());
+
+  inferTransitiveBindings(cs, existingTypes, inferredBindings);
+
+  inferDefaultTypes(cs, existingTypes);
+
+  // Adjust optionality of existing bindings based on presence of
+  // `ExpressibleByNilLiteral` requirement.
+  if (llvm::any_of(Protocols, [](Constraint *constraint) {
+        auto *protocol = constraint->getProtocol();
+        return protocol->isSpecificProtocol(
+            KnownProtocolKind::ExpressibleByNilLiteral);
+      })) {
+    for (auto &binding : Bindings) {
+      bool wrapInOptional = false;
+      if (binding.Kind == AllowedBindingKind::Supertypes) {
+        auto type = binding.BindingType->getRValueType();
+        // If the type doesn't conform to ExpressibleByNilLiteral,
+        // produce an optional of that type as a potential binding. We
+        // overwrite the binding in place because the non-optional type
+        // will fail to type-check against the nil-literal conformance.
+        bool conformsToExprByNilLiteral = false;
+        if (auto *nominalBindingDecl = type->getAnyNominal()) {
+          SmallVector<ProtocolConformance *, 2> conformances;
+          conformsToExprByNilLiteral = nominalBindingDecl->lookupConformance(
+              cs.DC->getParentModule(),
+              cs.getASTContext().getProtocol(
+                  KnownProtocolKind::ExpressibleByNilLiteral),
+              conformances);
+        }
+        wrapInOptional = !conformsToExprByNilLiteral;
+      } else if (binding.isDefaultableBinding() &&
+                 binding.BindingType->isAny()) {
+        wrapInOptional = true;
+      }
+
+      if (wrapInOptional)
+        binding.BindingType = OptionalType::get(binding.BindingType);
+    }
+  }
+
+  // If there are no bindings, typeVar may be a hole.
+  if (cs.shouldAttemptFixes() && Bindings.empty() &&
+      TypeVar->getImpl().canBindToHole()) {
+    IsHole = true;
+    // If the base of the unresolved member reference like `.foo`
+    // couldn't be resolved we'd want to bind it to a hole at the
+    // very last moment possible, just like generic parameters.
+    auto *locator = TypeVar->getImpl().getLocator();
+    if (locator->isLastElement<LocatorPathElt::MemberRefBase>())
+      PotentiallyIncomplete = true;
+
+    addPotentialBinding(PotentialBinding::forHole(cs.getASTContext(), locator));
+  }
+
+  // Let's always consider `Any` to be a last resort binding because
+  // it's always better to infer concrete type and erase it if required
+  // by the context.
+  if (Bindings.size() > 1) {
+    auto anyType = llvm::find_if(Bindings, [](const PotentialBinding &binding) {
+      return binding.BindingType->isAny() && !binding.isDefaultableBinding();
+    });
+
+    if (anyType != Bindings.end()) {
+      std::rotate(Bindings.begin(), anyType + 1, Bindings.end());
```

Comment:
If the order of `Bindings` is important, why is a rotation okay here? Why isn't this moving _only_ the `Any` binding to the end?

Reply:
`std::rotate` is only moving `Any` to the end, it should replace its current position with next in the iterator and move it to after the .end(). Maybe I misunderstood that API though...

Reply:
Ah I think I understand what is the problem, here, I'd need some other way to only move one element...

Reply:
Actually, it looks like what I need to do is `std::rotate(anyType, anyType + 1, Bindings.end());`